### DISCUSSION TITLE
How to change rules without container restart?

### DISCUSSION THREAD

**Discussion by exander77** [shinebayar-g/ufw-docker-automated#59] [Q&A]
Docker doesn't seem to be able to dynamically change labels.
I have large mysql databases that need to run without downtime.
I don't see any way to reload rules except for adding them to compose and applying them also manually using ufw directly.

**Comment by shinebayar-g** [comment]
IIRC you can't change docker published ports dynamically either right?

**Reply by exander77** [reply]
I am not sure how is that related or relevant.

**Reply by shinebayar-g** [reply]
Well yeah it's not totally related lol nvm that.
So you want to implement ufw-docker rules to already running containers? I believe labels require container restart. So best possible action would be something like reading external files that describes the rules and attach them. This would be a new feature implementation.

But if you're not going to restart your container you can just add the rules manually right? Do you want to connect it to ufw-docker-automated to make it automated in case container ever restarts?

**Reply by exander77** [reply]
Yes, you can add rules into docker compose file and also manually, but there is a risk that you do something differently, and it will break after restart. Adding it to compose and applying it would be safer.

And yes, you can't change open ports or labels on the running container which I personally think is oversight (because there isn't technical reason). But I kinda understand it for ports, as if there isn't something running on that port in the first place, you need to rebuild the container anyway. Ports are usually set one time.

On the other hand firewall rules can change a lot.

**Reply by shinebayar-g** [reply]
So do you have anything on your mind? I'm thinking maybe we could add a configuration file for ufw-docker-automated. Configuration file will be similar to container label based rules. But since you can't update your containers' labels without restarting them, we have to somehow let ufw-docker-automated know that we want to update existing containers.

One way to do that would be using container name. As I believe they're unique per docker host.

```
❯ docker run -d --name test nginx:alpine
224d7f3a14074e48ffba10de77ab140a95d02c61b43d8ef62b466e4b13761e99

❯ docker run -d --name test nginx:alpine
docker: Error response from daemon: Conflict. The container name "/test" is already in use by container "224d7f3a14074e48ffba10de77ab140a95d02c61b43d8ef62b466e4b13761e99". You have to remove (or rename) that container to be able to reuse that name.
See 'docker run --help'.
```

So it's guaranteed that each container will have unique name, so we can be sure that ufw rules won't conflict. Wdyt?

**Reply by exander77** [reply]
I was thinking more like a similar behavior to `docker-compose`.
Running something like:
```
ufw-docker-automated update
```
Would read docker compose file and run `docker compose ps`, if container is running and there are new rules in lables, then update ufw.
For running docker without compose
For docker without compose that's tough.

**Reply by shinebayar-g** [reply]
I see. But that's still not enough. For example, ufw-docker-automated only filters out containers with UFW_MANAGED label attached from the docker event stream. So even if we read docker-compose file, we're still not going to find the UFW_MANAGED label from existing containers... So ¯\_(ツ)_/¯

**Reply by exander77** [reply]
It's not perfect solution. Ideally, combine it with output of `docker compose ps`.
I looked if there isn't part of docker metadata that can be updated without restart, but nope.

**Reply by exander77** [reply]
Rethinking it, maybe the docker metadata should not contain actual rules, but paths:
```
UFW_ALLOW_FROM_RULES: mysql-in
UFW_ALLOW_TO_RULES: mysql-out
```
And docker automated could hook on the files through inotify a reload the rules on file change.

Without path, the rules would be taken from `/etc/ufw-docker-automated/rules.d/mysql-in.rule`, etc.
This way, rules could be also shared between several containers. Absolute path could also be supported or make `UFW_RULES_PATH` variable.

**Reply by exander77** [reply]
You could also set if rules will apply right away, or if you will have pending changes you need to apply:
```
ufw-dockers-automated pending
ufw-dockers-automated pending apply
```
Just a thought.

**Reply by exander77** [reply]
You could also merge rules into single yaml files to cover from, to, etc.:
```yaml
enabled: true
denyOut: true
from:
  - 172.10.50.32:8080,192.168.3.0/24:8080,10.10.0.50/32:8080--LAN
  - 172.10.50.32,192.168.3.0/24,10.10.0.50/32:8080--LAN
  - 10.10.0.50/32:8090-8100--LAN
  - 10.10.0.50/32:8020,8030,8040--LAN
  - 10.10.0.51,10.10.0.52,10.10.0.53:8020,8030,8040--LAN
  - 10.10.0.50-10.10.0.60:8020,8030,8040--LAN
  - :8888--EveryWhere
to:
  - 8.8.8.8:53--GoogleDNS
  - 1.1.1.0/24:53--CloudflareDNS
  - 192.168.10.24:8080--LAN
```
So you can combine various things.

**Reply by shinebayar-g** [reply]
Yup I like the idea. So it'll allow dynamic updates without necessarily requiring container restarts as all the metadata are immutable. One thing though,

If we do it this way
`UFW_ALLOW_FROM_RULES: mysql-in` is it a container label?

If yes, it'll still require your container to restart to change/apply different rules.

Or we can lookup containers by name from the configuration file itself. So no metadata is required on containers.

**Reply by exander77** [reply]
No, it is meant as an identification of the rule file: `/etc/ufw-docker-automated/rules.d/mysql-in.rule`
Something like that, but that's just initial brainstorming, I can also imagine writing absolute path there:
```
UFW_ALLOW_FROM_RULES: /etc/ufw-docker-automated/rules.d/mysql-in.rule
```
Or just file that searched in a specific location:
```
UFW_ALLOW_FROM_RULES: mysql-in.rule
UFW_ALLOW_FROM_RULES: mysql-in
```
I am not sure what would be preferable. I personally don't like absolute path, and also doesn't like forcing it to be in `/etc` for non root users.
So maybe:
```
UFW_RULES_PATH: /etc/ufw-docker-automated/rules.d
UFW_RULES_PATH: ~/.ufw-docker-automated/rules
```
And then just write:
```
UFW_ALLOW_FROM_RULES: mysql-in.rule
UFW_ALLOW_FROM_RULES: mysql-in
```
Where `mysql-in` corresponds with filename which is looked up in `UFW_RULES_PATH`.

**Reply by exander77** [reply]
It would be great to be able to store rule files relative to `docker-compose.yml` location, but I don't see how to get that path from running container.

**Reply by exander77** [reply]
But I can always do:
```
UFW_RULES_PATH: ${PWD}/ufw
```
And I can do similar for running container directly without compose.

**Reply by exander77** [reply]
Btw, I would be willing to lend a hand with implementation as it sounds very useful.

**Reply by exander77** [reply]
It could actually read compose working dir directly from events.
```
com.docker.compose.project.working_dir
```

**Reply by shinebayar-g** [reply]
You could give it a shot. But I'd like to request to make configuration file completely separate from containers. No labels required on containers.

**Reply by exander77** [reply]
Do you mean completely without labels like these:
```
UFW_MANAGED=true
UFW_RULES_PATH: ~/.ufw-docker-automated/rules
UFW_ALLOW_FROM_RULES: mysql-in
```
?
Or just without the actual rules that will be inside files?

**Reply by shinebayar-g** [reply]
Yeah completely without labels. Because I realized that if we need to update the labels to apply some changes it's always going to require re-deploying the container and that's no good IMO.

**Reply by exander77** [reply]
Yeah, I don't like either. But how connect rules with a container?

For compose I can do this, take: `com.docker.compose.project.working_dir` and join it with `ufw.rules` or `ufw.yaml` and expect the file to be in the project directory. But for directly run docker images? Use container names? Any ideas?

**Reply by shinebayar-g** [reply]
> But how connect rules with a container?

By container name. Gotta careful with docker-compose containers, as they have slightly different naming convention.

Something like this as you suggested above works for me `/etc/ufw-docker-automated/rules.d/..`.
